README.md
PHP ICO - The PHP ICO Generator
===============================

PHP ICO provides an easy-to-use PHP class that generates valid [ICO files](http://en.wikipedia.org/wiki/ICO_%28file_format%29). Note that these are not simply BMP, GIF, or PNG formatted images with an extension of ico, the images generated by this class are fully valid ICO-formatted image files.

The class uses the GD library for reading an image from the source file and uses pure PHP for generating the ICO file format. In theory, any image format that GD can read can be used as a source image to generate the ICO file. I tested this library with JPEG, GIF, and PNG images with great results. If an animated GIF is supplied, the first frame will be used to generate the ICO image.

The PHP ICO library is available on Composer via Packagist at `chrisjean/php-ico`.

ICO Format Details
------------------

The primary goal of creating this class was to have a simple-to-use and reliable method of generating ICO files for use as [favicons](http://en.wikipedia.org/wiki/Favicon) in websites. This goal drove much of the development and is the reason for some of the limitations in the resulting functionality.

ICO files support two different ways of encoding the actual image data: the [Windows BMP](http://en.wikipedia.org/wiki/BMP_file_format) format and the [PNG](http://en.wikipedia.org/wiki/Portable_Network_Graphics) format. Since support for the PNG format was introduced in Windows Vista and both older and newer versions of Windows support the BMP enocoding, the BMP encoding method was used.

Images are encoded using 32 bits per pixel. This allows for alpha channel information in the resulting images. Support for 32 bit images was added in Windows XP. This means that older versions of Windows are not likely to display the ICO images properly, if at all. The generated images have not been tested on versions of Windows predating XP.

Usage
-----

The following is a very basic example of using the PHP ICO library:

```php
require( dirname( __FILE__ ) . '/class-php-ico.php' );

$source = dirname( __FILE__ ) . '/example.gif';
$destination = dirname( __FILE__ ) . '/example.ico';

$ico_lib = new PHP_ICO( $source );
$ico_lib->save_ico( $destination );
````

It takes a source file named `example.gif` and produce an output ICO file named `example.ico`. `example.ico` will contain a single image that is the same size as the source image.

You can also render the gif directly on the browser without saving it on the server:

```php
require( dirname( __FILE__ ) . '/class-php-ico.php' );

$source = dirname( __FILE__ ) . '/example.gif';

$ico_lib = new PHP_ICO( $source );
$ico_lib->render_ico();
````

The ICO file format is capable of holding multiple images, each of a different size. The PHP ICO library opens up this feature of the ICO format as shown in the following example:

```php
require( dirname( __FILE__ ) . '/class-php-ico.php' );

$source = dirname( __FILE__ ) . '/example.gif';
$destination = dirname( __FILE__ ) . '/example.ico';

$ico_lib = new PHP_ICO( $source, array( array( 32, 32 ), array( 64, 64 ) ) );
$ico_lib->save_ico( $destination );
```

As with the previous example, this example produces `example.ico` from the `example.gif` source file. In this example, sizes were passed to the constructor that result in the `example.ico` file containing two images: one that is 32x32 and one that is 64x64. Since the same source image is used for each of the contained images, each contained image is simply the source image scaled to the new dimensions.

Using different source images for the different sizes contained inside an ICO file can create much better results. For instance, you can use a high-quality image for higher-resolution images and use smaller images that are tailored for their specific dimensions for the smaller sizes. The following example shows how the PHP ICO library can be used to create such ICO files:

```php
require( dirname( __FILE__ ) . '/class-php-ico.php' );

$destination = dirname( __FILE__ ) . '/example.ico';

$ico_lib = new PHP_ICO();

$ico_lib->add_image( dirname( __FILE__ ) . '/example-small.gif', array( array( 16, 16 ), array( 24, 24 ), array( 32, 32 ) ) );
$ico_lib->add_image( dirname( __FILE__ ) . '/example-medium.gif', array( array( 48, 48 ), array( 96, 96 ) ) );
$ico_lib->add_image( dirname( __FILE__ ) . '/example-large.gif', array( array( 128, 128 ) ) );

$ico_lib->save_ico( $destination );
```

This example creates a single ICO file named `example.ico` just as the previous examples. The difference is that this example used multiple source images to generate the final ICO images. The final ICO image contains a total of six images: 16x16, 24x24, and 32x32 images from `example-small.gif`; 48x48 and 96x96 images from `example-medium.gif`; and a 128x128 image from `example-large.gif`.

By using this feature of supplying multiple source images with specific sizes for each, you can generate ICO files that have very high-quality images at each supplied resolution.

Since the PHP ICO library was created to generate favicon files, the following example shows how to generate a favicon ICO file with all the needed dimensions from a single source image:

```php
require( dirname( __FILE__ ) . '/class-php-ico.php' );

$source = dirname( __FILE__ ) . '/example.gif';
$destination = dirname( __FILE__ ) . '/example.ico';

$sizes = array(
	array( 16, 16 ),
	array( 24, 24 ),
	array( 32, 32 ),
	array( 48, 48 ),
);

$ico_lib = new PHP_ICO( $source, $sizes );
$ico_lib->save_ico( $destination );
```

I've found that the 16x16, 24x24, 32x32, and 48x48 image sizes cover all the sizes that browsers and Windows will try to use a favicon image for. The different sizes come from using the favicon for various browser icons, bookmarks, and various other places.

Thanks
------

I'd like to thank [iThemes](http://ithemes.com) for making this project possible. This code was originally developed to add easy-to-use favicon support to the [Builder theme](http://ithemes.com/purchase/builder-theme/). I asked [Cory](http://corymiller.tv/), owner of iThemes and my boss, if I could share the final code. He gave me the green light. Win for everyone. :)

Thanks iThemes. Thanks Cory.


changelog
1.0.0 - 2011-08-04
	Initial version.
1.0.1 - 2012-10-01
	Added requirements checks.
1.0.2 - 2013-01-07
	Added bug fix that caused source images with transparency to have artifacts when output at the same dimensions as the source.
1.0.3 - 2016-07-22
	Updated class constructor to __construct() as PHP4-style constructors are now deprecated. Props @nuxodin.
	Removed trailing whitespace from blank lines.
1.0.4 - 2016-09-27
	Updated version to get composer to track properly.
2.0.0 - 2023-08-02
    Comptability PHP 8


test/test-converter.php
<?php
use Ossobuffo\PhpIco\IcoConverter;

require_once __DIR__ . '/../src/IcoConverter.php';

$source = __DIR__ . '/input.png';
$source_small = __DIR__ . '/input-32x32.gif';

// Create an enormous 250x250 .ico file
$destination = __DIR__ . '/output1.ico';
$converter = new IcoConverter($source);
$converter->saveIco($destination);

// Create an .ico file with two sized images from the same source.
$destination = __DIR__ . '/output2.ico';
$dimensions = [[16, 16], [32, 32]];
$converter = new IcoConverter($source, $dimensions);
$converter->saveIco($destination);

// Create an .ico file with two sized images from different sources.
$destination = __DIR__ . '/output3.ico';
$converter = new IcoConverter();
$converter->addImage($source_small, [[32, 32]]);
$converter->addImage($source, [[64, 64]]);
$converter->saveIco($destination);


tests/OutputIcoTest.php
<?php
declare (strict_types=1);

namespace PHP_ICO\Tests;

use PHP_ICO;
use PHPUnit_Framework_TestCase;

/**
 * Class EasyDBTest
 * @package ParagonIE\EasyDB\Tests
 */
class OutputIcoTest extends PHPUnit_Framework_TestCase
{

    public function goodAddImageSingleProvider()
    {
        return array(
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.gif',
                array(),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.jpg',
                array(),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.png',
                array(),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.gif',
                array(16,16),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.jpg',
                array(16, 16),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.png',
                array(16, 16),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.gif',
                array(
                    array(16,16),
                ),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.jpg',
                array(
                    array(16,16),
                ),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.png',
                array(
                    array(16,16),
                ),
            ),
        );
    }

    public function badAddImageSingleProvider_invalidFile()
    {
        return array(
            array(
                null,
                array(),
            ),
            array(
                false,
                array(),
            ),
            array(
                true,
                array(),
            ),
            array(
                1,
                array(),
            ),
            array(
                '',
                array(),
            ),
            array(
                __DIR__ . DIRECTORY_SEPARATOR . 'test-ico-1.xcf',
                array(),
            ),
        );
    }

    public function badSaveIcoProvider_GetIcoDataReturnsFalse()
    {
        return array(
            array(
                array(),
            ),
        );
    }

    /**
    * @dataProvider goodAddImageSingleProvider
    */
    public function testConstructorOnly($file, $sizes) {
        $this->assertTrue(is_file($file));
        $this->assertTrue(is_readable($file));
        $this->assertTrue(is_file($file . '.ico'));
        $this->assertTrue(is_readable($file . '.ico'));

        $ico = new PHP_ICO($file, $sizes);
        $outputToHere = tempnam(sys_get_temp_dir(), 'PHP_ICO_tests');
        $this->assertTrue($ico->save_ico($outputToHere));
        $this->assertSame(sha1_file($file . '.ico'), sha1_file($outputToHere));
        unlink($outputToHere);
    }

    /**
    * @dataProvider badAddImageSingleProvider_invalidFile
    */
    public function testAddImageBadFiles($file, $sizes) {
        $ico = new PHP_ICO();
        $this->assertFalse($ico->add_image($file, $sizes));
    }

    /**
    * @dataProvider badSaveIcoProvider_GetIcoDataReturnsFalse
    */
    public function testSaveIcoBadData($arrayOfFilesAndSizes)
    {
        $ico = new PHP_ICO();
        foreach ($arrayOfFilesAndSizes as $file => $sizes)
        {
            $ico->add_image($file, $sizes);
        }
        $outputToHere = tempnam(sys_get_temp_dir(), 'PHP_ICO_tests');
        $this->assertFalse($ico->save_ico($outputToHere));
        unlink($outputToHere);
    }
}


class-php-ico.php
<?php

/*
Copyright 2011-2016 Chris Jean & iThemes
Licensed under GPLv2 or above

Version 1.0.4
*/

class PHP_ICO {
	/**
	 * Images in the BMP format.
	 *
	 * @var array
	 * @access private
	 */
	private $_images = [];

	/**
	 * Flag to tell if the required functions exist.
	 *
	 * @var boolean
	 * @access private
	 */
	private $_has_requirements = false;


	/**
	 * Constructor - Create a new ICO generator.
	 *
	 * If the constructor is not passed a file, a file will need to be supplied using the {@link PHP_ICO::add_image}
	 * function in order to generate an ICO file.
	 *
	 * @param string $file Optional. Path to the source image file.
	 * @param array $sizes Optional. An array of sizes (each size is an array with a width and height) that the source image should be rendered at in the generated ICO file. If sizes are not supplied, the size of the source image will be used.
	 */
	public function __construct( $file = false, array $sizes = [] ) {
		$required_functions = [
			'getimagesize',
			'imagecreatefromstring',
			'imagecreatetruecolor',
			'imagecolortransparent',
			'imagecolorallocatealpha',
			'imagealphablending',
			'imagesavealpha',
			'imagesx',
			'imagesy',
			'imagecopyresampled'
		];

		foreach ( $required_functions as $function ) {
			if ( ! function_exists( $function ) ) {
				trigger_error( "The PHP_ICO class was unable to find the $function function, which is part of the GD library. Ensure that the system has the GD library installed and that PHP has access to it through a PHP interface, such as PHP's GD module. Since this function was not found, the library will be unable to create ICO files." );
				return;
			}
		}

		$this->_has_requirements = true;


		if ( false != $file )
			$this->add_image( $file, $sizes );
	}

	/**
	 * Add an image to the generator.
	 *
	 * This function adds a source image to the generator. It serves two main purposes: add a source image if one was
	 * not supplied to the constructor and to add additional source images so that different images can be supplied for
	 * different sized images in the resulting ICO file. For instance, a small source image can be used for the small
	 * resolutions while a larger source image can be used for large resolutions.
	 *
	 * @param string $file Path to the source image file.
	 * @param array $sizes Optional. An array of sizes (each size is an array with a width and height) that the source image should be rendered at in the generated ICO file. If sizes are not supplied, the size of the source image will be used.
	 * @return boolean true on success and false on failure.
	 */
	public function add_image( string $file, array $sizes = [] ):bool {
		if ( ! $this->_has_requirements )
			return false;

		if ( false === ( $im = $this->_load_image_file( $file ) ) )
			return false;


		if ( empty( $sizes ) )
			$sizes = array( imagesx( $im ), imagesy( $im ) );

		// If just a single size was passed, put it in array.
		if ( ! is_array( $sizes[0] ) )
			$sizes = array( $sizes );

		foreach ( (array) $sizes as $size ) {
			list( $width, $height ) = $size;

			$new_im = imagecreatetruecolor( $width, $height );

			imagecolortransparent( $new_im, imagecolorallocatealpha( $new_im, 0, 0, 0, 127 ) );
			imagealphablending( $new_im, false );
			imagesavealpha( $new_im, true );

			$source_width = imagesx( $im );
			$source_height = imagesy( $im );

			if ( false === imagecopyresampled( $new_im, $im, 0, 0, 0, 0, $width, $height, $source_width, $source_height ) )
				continue;

			$this->_add_image_data( $new_im );
		}

		return true;
	}

	/**
	 * Write the ICO file data to a file path.
	 *
	 * @param string $file Path to save the ICO file data into.
	 * @return boolean true on success and false on failure.
	 */
	public function save_ico(string $file ) :bool {
		if ( ! $this->_has_requirements )
			return false;

		if ( false === ( $data = $this->_get_ico_data() ) )
			return false;

		if ( false === ( $fh = fopen( $file, 'w' ) ) )
			return false;

		if ( false === ( fwrite( $fh, $data ) ) ) {
			fclose( $fh );
			return false;
		}

		fclose( $fh );

		return true;
	}
    /**
     * Display the ICO file data.
     *
     * @return boolean true on success and false on failure.
     */
    public function render_ico():bool {
        if ( ! $this->_has_requirements )
            return false;

        if ( false === ( $data = $this->_get_ico_data() ) )
            return false;

        header('Content-Type: image/x-icon');
        echo $data;

        return true;
    }
	/**
	 * Generate the final ICO data by creating a file header and adding the image data.
	 *
	 * @access private
	 */
	private function _get_ico_data() {
		if ( ! is_array( $this->_images ) || empty( $this->_images ) )
			return false;


		$data = pack( 'vvv', 0, 1, count( $this->_images ) );
		$pixel_data = '';

		$icon_dir_entry_size = 16;

		$offset = 6 + ( $icon_dir_entry_size * count( $this->_images ) );

		foreach ( $this->_images as $image ) {
			$data .= pack( 'CCCCvvVV', $image['width'], $image['height'], $image['color_palette_colors'], 0, 1, $image['bits_per_pixel'], $image['size'], $offset );
			$pixel_data .= $image['data'];

			$offset += $image['size'];
		}

		$data .= $pixel_data;
		unset( $pixel_data );


		return $data;
	}

	/**
	 * Take a GD image resource and change it into a raw BMP format.
	 *
	 * @access private
	 */
	private function _add_image_data( $im ) {
		$width = imagesx( $im );
		$height = imagesy( $im );


		$pixel_data = array();

		$opacity_data = array();
		$current_opacity_val = 0;

		for ( $y = $height - 1; $y >= 0; $y-- ) {
			for ( $x = 0; $x < $width; $x++ ) {
				$color = imagecolorat( $im, $x, $y );

				$alpha = ( $color & 0x7F000000 ) >> 24;
				//$alpha = ( 1 - ( $alpha / 127 ) ) * 255;
                $alpha = round( ( 1 - ( $alpha / 127 ) ) * 255);

				$color &= 0xFFFFFF;
				$color |= 0xFF000000 & ( $alpha << 24 );

				$pixel_data[] = $color;


				$opacity = ( $alpha <= 127 ) ? 1 : 0;

				$current_opacity_val = ( $current_opacity_val << 1 ) | $opacity;

				if ( ( ( $x + 1 ) % 32 ) == 0 ) {
					$opacity_data[] = $current_opacity_val;
					$current_opacity_val = 0;
				}
			}

			if ( ( $x % 32 ) > 0 ) {
				while ( ( $x++ % 32 ) > 0 )
					$current_opacity_val = $current_opacity_val << 1;

				$opacity_data[] = $current_opacity_val;
				$current_opacity_val = 0;
			}
		}

		$image_header_size = 40;
		$color_mask_size = $width * $height * 4;
		$opacity_mask_size = ( ceil( $width / 32 ) * 4 ) * $height;


		$data = pack( 'VVVvvVVVVVV', 40, $width, ( $height * 2 ), 1, 32, 0, 0, 0, 0, 0, 0 );

		foreach ( $pixel_data as $color )
			$data .= pack( 'V', $color );

		foreach ( $opacity_data as $opacity )
			$data .= pack( 'N', $opacity );


		$image = [
			'width'                => $width,
			'height'               => $height,
			'color_palette_colors' => 0,
			'bits_per_pixel'       => 32,
			'size'                 => $image_header_size + $color_mask_size + $opacity_mask_size,
			'data'                 => $data,
		];

		$this->_images[] = $image;
	}

	/**
	 * Read in the source image file and convert it into a GD image resource.
	 *
	 * @access private
	 */
	private function _load_image_file( $file ) {
		// Run a cheap check to verify that it is an image file.
		if ( false === ( $size = getimagesize( $file ) ) )
			return false;

		if ( false === ( $file_data = file_get_contents( $file ) ) )
			return false;

		if ( false === ( $im = imagecreatefromstring( $file_data ) ) )
			return false;

		unset( $file_data );


		return $im;
	}
}


src/IcoConverter.php
<?php

namespace Ossobuffo\PhpIco;

/*
Copyright 2011-2013 Chris Jean & iThemes
Licensed under GPLv2 or above
*/

class IcoConverter
{
    /**
     * Images in the PNG format.
     *
     * @var array
     */
    private $images = [];

    /**
     * Size of BMP image header.
     */
    const IMAGE_HEADER_SIZE = 40;

    /**
     * Size of icon directory entry in the BMP index.
     */
    const ICON_DIR_ENTRY_SIZE = 16;

    /**
     * Constructor - Create a new ICO generator.
     *
     * If the constructor is not passed a file, a file will need to be supplied
     * using the addImage() function in order to generate an ICO file.
     *
     * @param string|bool $file
     *      Optional. Path to the source image file.
     * @param array $sizes
     *      Optional. An array of sizes (each size is an array with a width and
     *      height) that the source image should be rendered at in the
     *      generated ICO file. If sizes are not supplied, the size of the
     *      source image will be used.
     */
    public function __construct($file = false, array $sizes = [])
    {
        if (false != $file) {
            $this->addImage($file, $sizes);
        }
    }

    /**
     * Add an image to the generator.
     *
     * This function adds a source image to the generator. It serves two main
     * purposes: add a source image if one was not supplied to the constructor
     * and to add additional source images so that different images can be
     * supplied for different sized images in the resulting ICO file. For
     * instance, a small source image can be used for the small resolutions
     * while a larger source image can be used for large resolutions.
     *
     * @param string $file
     *      Path to the source image file.
     * @param array $sizes
     *      Optional. An array of sizes (each size is an array with a width
     *      and height) that the source image should be rendered at in the
     *      generated ICO file. If sizes are not supplied, the size of the
     *      source image will be used.
     *
     * @return boolean
     *      true on success and false on failure.
     */
    public function addImage($file, $sizes = [])
    {
        if (false === ($image = $this->loadImageFile($file))) {
            return false;
        }

        if (empty($sizes)) {
            $sizes = [imagesx($image), imagesy($image)];
        }

        // If just a single size was passed, put it in array.
        if (!is_array($sizes[0])) {
            $sizes = [$sizes];
        }

        foreach ((array)$sizes as $size) {
            list($width, $height) = $size;

            $newImage = imagecreatetruecolor($width, $height);

            imagecolortransparent($newImage, imagecolorallocatealpha($newImage, 0, 0, 0, 127));
            imagealphablending($newImage, false);
            imagesavealpha($newImage, true);

            $sourceWidth = imagesx($image);
            $sourceHeight = imagesy($image);

            $success = imagecopyresampled($newImage, $image, 0, 0, 0, 0, $width, $height, $sourceWidth, $sourceHeight);

            if (false === $success) {
                continue;
            }

            $this->addImageData($newImage);
        }

        return true;
    }

    /**
     * Write the ICO file data to a file path.
     *
     * @param string $file
     *      Path to save the ICO file data into.
     *
     * @return boolean
     *      true on success and false on failure.
     */
    public function saveIco($file)
    {
        if (false === ($data = $this->getIcoData())) {
            return false;
        }

        if (false === ($fh = fopen($file, 'w'))) {
            return false;
        }

        if (false === (fwrite($fh, $data))) {
            fclose($fh);
            return false;
        }

        fclose($fh);
        return true;
    }

    /**
     * Generate final ICO data by creating a file header & adding image data.
     */
    private function getIcoData()
    {
        if (!is_array($this->images) || empty($this->images)) {
            return false;
        }

        $data = pack('vvv', 0, 1, count($this->images));
        $pixelData = '';

        $offset = 6 + (self::ICON_DIR_ENTRY_SIZE * count($this->images));

        foreach ($this->images as $image) {
            $data .= pack(
                'CCCCvvVV',
                $image['width'],
                $image['height'],
                $image['color_palette_colors'],
                0,
                1,
                $image['bits_per_pixel'],
                $image['size'],
                $offset
            );
            $pixelData .= $image['data'];

            $offset += $image['size'];
        }

        $data .= $pixelData;
        unset($pixelData);
        return $data;
    }

    /**
     * Take a GD image resource and change it into a raw BMP format.
     *
     * The BMP data is appended to the $this->images array.
     *
     * @param resource $image
     *      The GD image to be changed into BMP.
     */
    private function addImageData($image)
    {
        $width = imagesx($image);
        $height = imagesy($image);

        $pixelData = [];

        $opacityData = [];
        $currentOpacityVal = 0;

        for ($y = $height - 1; $y >= 0; $y--) {
            for ($x = 0; $x < $width; $x++) {
                $color = imagecolorat($image, $x, $y);

                $alpha = ($color & 0x7F000000) >> 24;
                $alpha = (1 - ($alpha / 127)) * 255;

                $color &= 0xFFFFFF;
                $color |= 0xFF000000 & ($alpha << 24);

                $pixelData[] = $color;


                $opacity = ($alpha <= 127) ? 1 : 0;

                $currentOpacityVal = ($currentOpacityVal << 1) | $opacity;

                if ((($x + 1) % 32) == 0) {
                    $opacityData[] = $currentOpacityVal;
                    $currentOpacityVal = 0;
                }
            }

            if (($x % 32) > 0) {
                while (($x++ % 32) > 0) {
                    $currentOpacityVal = $currentOpacityVal << 1;
                }

                $opacityData[] = $currentOpacityVal;
                $currentOpacityVal = 0;
            }
        }

        $colorMaskSize = $width * $height * 4;
        $opacityMaskSize = (ceil($width / 32) * 4) * $height;

        $data = pack('VVVvvVVVVVV', 40, $width, ($height * 2), 1, 32, 0, 0, 0, 0, 0, 0);

        foreach ($pixelData as $color) {
            $data .= pack('V', $color);
        }

        foreach ($opacityData as $opacity) {
            $data .= pack('N', $opacity);
        }

        $image = [
            'width' => $width,
            'height' => $height,
            'color_palette_colors' => 0,
            'bits_per_pixel' => 32,
            'size' => self::IMAGE_HEADER_SIZE + $colorMaskSize + $opacityMaskSize,
            'data' => $data,
        ];

        $this->images[] = $image;
    }

    /**
     * Read in the source image file and convert it into a GD image resource.
     *
     * @param string $file
     *      Name of an image file to be loaded.
     *
     * @return resource
     *      The GD image resource created from the image file.
     */
    private function loadImageFile($file)
    {
        // Run a cheap check to verify that it is an image file.
        if (false === ($size = getimagesize($file))) {
            return false;
        }

        if (false === ($fileData = file_get_contents($file))) {
            return false;
        }

        if (false === ($im = imagecreatefromstring($fileData))) {
            return false;
        }

        unset($fileData);
        return $im;
    }
}


src/PhpIco.php
<?php

namespace Chrisbliss18;

/*
Copyright 2011-2013 Chris Jean & iThemes
Licensed under GPLv2 or above

Version 1.0.2
*/

class PhpIco
{
    /**
     * Images in the BMP format.
     *
     * @var array
     */
    private $images = array();

    /**
     * Flag to tell if the required functions exist.
     *
     * @var boolean
     */
    private $hasRequirements = false;

    /**
     * Size of BMP image header.
     */
    const IMAGE_HEADER_SIZE = 40;

    /**
     * Size of icon directory entry in the BMP index.
     */
    const ICON_DIR_ENTRY_SIZE = 16;


    /**
     * Constructor - Create a new ICO generator.
     *
     * If the constructor is not passed a file, a file will need to be supplied
     * using the addImage() function in order to generate an ICO file.
     *
     * @param string|bool $file
     *      Optional. Path to the source image file.
     * @param array $sizes
     *      Optional. An array of sizes (each size is an array with a width and
     *      height) that the source image should be rendered at in the
     *      generated ICO file. If sizes are not supplied, the size of the
     *      source image will be used.
     */
    public function __construct($file = false, array $sizes = array())
    {
        $requiredFunctions = array(
            'getimagesize',
            'imagecreatefromstring',
            'imagecreatetruecolor',
            'imagecolortransparent',
            'imagecolorallocatealpha',
            'imagealphablending',
            'imagesavealpha',
            'imagesx',
            'imagesy',
            'imagecopyresampled',
        );

        foreach ($requiredFunctions as $function) {
            if (!function_exists($function)) {
                throw new \RuntimeException(
                    'The ' . __CLASS__ . " class was unable to find the $function function, which is part of the GD "
                    . 'library. Ensure that the system has the GD library installed and that PHP has access to it '
                    . 'through a PHP interface, such as PHP’s GD module. Since this function was not found, the '
                    . 'library will be unable to create ICO files.'
                );
            }
        }

        $this->hasRequirements = true;

        if (false != $file) {
            $this->addImage($file, $sizes);
        }
    }

    /**
     * Add an image to the generator.
     *
     * This function adds a source image to the generator. It serves two main
     * purposes: add a source image if one was not supplied to the constructor
     * and to add additional source images so that different images can be
     * supplied for different sized images in the resulting ICO file. For
     * instance, a small source image can be used for the small resolutions
     * while a larger source image can be used for large resolutions.
     *
     * @param string $file
     *      Path to the source image file.
     * @param array $sizes
     *      Optional. An array of sizes (each size is an array with a width
     *      and height) that the source image should be rendered at in the
     *      generated ICO file. If sizes are not supplied, the size of the
     *      source image will be used.
     *
     * @return boolean
     *      true on success and false on failure.
     */
    public function addImage($file, $sizes = array())
    {
        if (!$this->hasRequirements) {
            return false;
        }

        if (false === ($image = $this->loadImageFile($file))) {
            return false;
        }


        if (empty($sizes)) {
            $sizes = array(imagesx($image), imagesy($image));
        }

        // If just a single size was passed, put it in array.
        if (!is_array($sizes[0])) {
            $sizes = array($sizes);
        }

        foreach ((array)$sizes as $size) {
            list($width, $height) = $size;

            $newImage = imagecreatetruecolor($width, $height);

            imagecolortransparent($newImage, imagecolorallocatealpha($newImage, 0, 0, 0, 127));
            imagealphablending($newImage, false);
            imagesavealpha($newImage, true);

            $sourceWidth = imagesx($image);
            $sourceHeight = imagesy($image);

            $success = imagecopyresampled($newImage, $image, 0, 0, 0, 0, $width, $height, $sourceWidth, $sourceHeight);

            if (false === $success) {
                continue;
            }

            $this->addImageData($newImage);
        }

        return true;
    }

    /**
     * Write the ICO file data to a file path.
     *
     * @param string $file
     *      Path to save the ICO file data into.
     *
     * @return boolean
     *      true on success and false on failure.
     */
    public function saveIco($file)
    {
        if (!$this->hasRequirements) {
            return false;
        }

        if (false === ($data = $this->getIcoData())) {
            return false;
        }

        if (false === ($fh = fopen($file, 'w'))) {
            return false;
        }

        if (false === (fwrite($fh, $data))) {
            fclose($fh);
            return false;
        }

        fclose($fh);
        return true;
    }

    /**
     * Generate final ICO data by creating a file header & adding image data.
     */
    private function getIcoData()
    {
        if (!is_array($this->images) || empty($this->images)) {
            return false;
        }

        $data = pack('vvv', 0, 1, count($this->images));
        $pixelData = '';

        $offset = 6 + (self::ICON_DIR_ENTRY_SIZE * count($this->images));

        foreach ($this->images as $image) {
            $data .= pack(
                'CCCCvvVV',
                $image['width'],
                $image['height'],
                $image['color_palette_colors'],
                0,
                1,
                $image['bits_per_pixel'],
                $image['size'],
                $offset
            );
            $pixelData .= $image['data'];

            $offset += $image['size'];
        }

        $data .= $pixelData;
        unset($pixelData);
        return $data;
    }

    /**
     * Take a GD image resource and change it into a raw BMP format.
     *
     * The BMP data is appended to the $this->images array.
     *
     * @param resource $image
     *      The GD image to be changed into BMP.
     */
    private function addImageData($image)
    {
        $width = imagesx($image);
        $height = imagesy($image);

        $pixelData = array();

        $opacityData = array();
        $currentOpacityVal = 0;

        for ($y = $height - 1; $y >= 0; $y--) {
            for ($x = 0; $x < $width; $x++) {
                $color = imagecolorat($image, $x, $y);

                $alpha = ($color & 0x7F000000) >> 24;
                $alpha = (1 - ($alpha / 127)) * 255;

                $color &= 0xFFFFFF;
                $color |= 0xFF000000 & ($alpha << 24);

                $pixelData[] = $color;


                $opacity = ($alpha <= 127) ? 1 : 0;

                $currentOpacityVal = ($currentOpacityVal << 1) | $opacity;

                if ((($x + 1) % 32) == 0) {
                    $opacityData[] = $currentOpacityVal;
                    $currentOpacityVal = 0;
                }
            }

            if (($x % 32) > 0) {
                while (($x++ % 32) > 0) {
                    $currentOpacityVal = $currentOpacityVal << 1;
                }

                $opacityData[] = $currentOpacityVal;
                $currentOpacityVal = 0;
            }
        }

        $colorMaskSize = $width * $height * 4;
        $opacityMaskSize = (ceil($width / 32) * 4) * $height;

        $data = pack('VVVvvVVVVVV', 40, $width, ($height * 2), 1, 32, 0, 0, 0, 0, 0, 0);

        foreach ($pixelData as $color) {
            $data .= pack('V', $color);
        }

        foreach ($opacityData as $opacity) {
            $data .= pack('N', $opacity);
        }

        $image = array(
            'width' => $width,
            'height' => $height,
            'color_palette_colors' => 0,
            'bits_per_pixel' => 32,
            'size' => self::IMAGE_HEADER_SIZE + $colorMaskSize + $opacityMaskSize,
            'data' => $data,
        );

        $this->images[] = $image;
    }

    /**
     * Read in the source image file and convert it into a GD image resource.
     *
     * @param string $file
     *      Name of an image file to be loaded.
     *
     * @return resource
     *      The GD image resource created from the image file.
     */
    private function loadImageFile($file)
    {
        // Run a cheap check to verify that it is an image file.
        if (false === ($size = getimagesize($file))) {
            return false;
        }

        if (false === ($fileData = file_get_contents($file))) {
            return false;
        }

        if (false === ($im = imagecreatefromstring($fileData))) {
            return false;
        }

        unset($fileData);
        return $im;
    }
}
